netket/_vmc.py

```diff
@@ -197,34 +202,50 @@ def advance(self, n_steps=1):
             for _ in self._sampler.samples(self._n_discard):
                 pass
 
-            # Generate samples
+            # Generate samples and store them
             for i, sample in enumerate(self._sampler.samples(self._n_samples_node)):
-
-                # Store the current sample
                 self._samples[i] = sample
 
-                # Compute Log derivatives
-                self._der_logs[i] = self._machine.der_log(sample)
-
-            # Center the log derivatives
-            _subtract_mean(self._der_logs)
-
-            # Estimate energy
+            # Compute the local energy estimator and average Energy
             eloc, self._stats = self._get_mc_stats(self._ham)
 
-            # Estimate energy gradient
-            grad = _covariance_sv(eloc, self._der_logs, center_s=False)
-
             # Perform update
             if self._sr:
-                dp = _np.empty(self._npar, dtype=_np.complex128)
+                # When using the SR (Natural gradient) we need to have the full jacobian
+                # Computes the jacobian
+                for i, sample in enumerate(self._samples):
+                    self._der_logs[i] = self._machine.der_log(sample)
+
                 # flatten MC chain dimensions:
-                derlogs = self._der_logs.reshape(-1, self._der_logs.shape[-1])
-                self._sr.compute_update(derlogs, grad, dp)
-                derlogs = self._der_logs.reshape(
-                    self._n_samples_node, self._n_chains, self._npar
+                self._der_logs = self._der_logs.reshape(-1, self._npar)
+
+                # Center the local energy
+                eloc -= _mean(eloc)
+
+                # Center the log derivatives
+                self._der_logs -= _mean(self._der_logs, axis=0)
+
+                # Compute the gradient
+                self._grads = _np.conjugate(self._der_logs) * eloc.reshape(-1, 1)
+
+                grad = _mean(self._grads, axis=0)
+
+                dp = _np.empty(self._npar, dtype=_np.complex128)
+
+                self._sr.compute_update(self._der_logs, grad, dp)
+
+                self._der_logs = self._der_logs.reshape(
+                    self._n_samples_node, self._batch_size, self._npar
                 )
             else:
+                # Computing updates using the simple gradient
+                # Center the local energy
+                eloc -= _mean(eloc)
+
+                for x, eloc_x, grad_x in zip(self._samples, eloc, self._grads):
+                    self._machine.vector_jacobian_prod(x, eloc_x, grad_x)
+
+                grad = _mean(self._grads, axis=0) / float(self._batch_size)
```

Comment:
Why do you divide the result of `_mean` by the batch size? Shouldn't this be handled by mean?

Reply:
The issue is that vector_jacobian_prod compute just the product of the jacobian (i.e. O_k(x) where x is a batch of visible values) time a vector. Thus to get the average over the batch you need to divide by hand by the batch_size...